### PR TITLE
gh-105227: Add PyType_GetDict()

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
This compensates for static builtin types having `tp_dict` set to `NULL`.

* Issue: gh-105227

----
:books: Documentation preview :books:: https://cpython-previews--105747.org.readthedocs.build/

### CODE REVIEW

**Comment by ericsnowcurrently** [comment]
CC @The-Compiler. @tacaswell, @philthompson10

File: Doc/c-api/type.rst [hunk 1]
```diff
@@ -50,6 +50,22 @@ Type Objects
...
+   Return the type object's internal namespace, which is otherwise only
+   exposed via a read-only proxy (``cls.__dict__``).  This is a
+   replacement for accessing :c:member:`~PyTypeObject.tp_dict` directly.
+   The returned dictionary must be treated as read-only.
+
+   This function isn't intended for general use.  It's meant for
+   specific embedding and language-binding cases, where direct access
+   to the dict is necessary and indirect access (e.g. via the proxy)
+   isn't adequate.  Extension modules may continue to use ``tp_dict``,
+   directly or indirectly, when setting up their own types.
```
Comment (erlend-aasland) [inline Doc/c-api/type.rst]: Maybe elevate this paragraph to a `.. note:`, or maybe a `.. warning:`?

File: Misc/NEWS.d/next/C API/2023-06-13-14-24-55.gh-issue-105227.HDL9aF.rst [hunk 2]
```diff
@@ -0,0 +1,5 @@
+The new :c:func:`PyType_GetDict` provides the dictionary for the given type
+object that is normally exposed by ``cls.__dict__``.  Normally it's
+sufficient to use :c:member:`~PyTypeObject.tp_dict`, but for the static
+builtin types ``tp_dict`` is now always ``NULL``.  ``PyType_GetDict()``
```
Comment (erlend-aasland) [inline Misc/NEWS.d/next/C API/2023-06-13-14-24-55.gh-issue-105227.HDL9aF.rst]: Formatting nit:

```suggestion
builtin types :c:member:`!tp_dict` is now always ``NULL``.  :c:func:`!PyType_GetDict()`
```

File: Doc/c-api/typeobj.rst [hunk 3]
```diff
@@ -157,6 +157,9 @@ Quick Reference
    **<>**: Names in angle brackets should be initially set to ``NULL`` and
    treated as read-only.
 
+   **<<>>**: Names in double angle brackets should be initially set to
+   ``NULL`` and treated as read-only after initialization.
```
Comment (encukou) [inline Doc/c-api/typeobj.rst]: Only marking up `tp_dict` implies it's the only one in this category. At least `tp_bases` & `tp_mro` should be marked too.

**Comment by encukou** [comment]
@Yhg1s needs to approve backporting to 3.12

File: Doc/c-api/type.rst [hunk 4]
```diff
@@ -50,6 +50,22 @@ Type Objects
...
 
 
+.. c:function:: PyObject* PyType_GetDict(PyTypeObject* type)
+
+   Return the type object's internal namespace, which is otherwise only
+   exposed via a read-only proxy (``cls.__dict__``).  This is a
+   replacement for accessing :c:member:`~PyTypeObject.tp_dict` directly.
+   The returned dictionary must be treated as read-only.
+
+   This function isn't intended for general use.  It's meant for
```
Comment (encukou) [inline Doc/c-api/type.rst]: If it's not meant for general use, would it be better to add it as unstable API?

**Review by erlend-aasland — COMMENTED** [review]
Can you update `Doc/data/refcounts.dat`, so we get a _"Return value: New reference"_ markup in the docs?

File: Doc/c-api/type.rst (hunk 4 above)
Comment (Yhg1s) [reply Doc/c-api/type.rst]: To be honest, I think it makes more sense to say this _is_ for general use.

**Review by Yhg1s — COMMENTED** [review]
Approving the new API as RM: 👍

**Comment by philthompson10** [comment]
PyQt now uses this call. Strictly speaking it could do the C equivalent of 'cls.__dict__' but the call is much nicer.

**Comment by ericsnowcurrently** [comment]
I'm not available enough to merge this for a few weeks, so anyone that feels comfortable doing so is welcome to in the meantime. 😄

**Comment by The-Compiler** [comment]
Any updates on this? This currently blocks me (with qutebrowser, via PyQt) from testing Python 3.12 on CI.

**Comment by encukou** [comment]
Oh dear, I didn't realis a few weeks is all the time until Beta. And I was busy so it didn't reach the top of my TODO list.
I'll push some tests and a doc update.

**Comment by The-Compiler** [comment]
Oh whoops, I wasn't even realizing how close to the Beta 4 I was commenting (apparently it was [scheduled](https://peps.python.org/pep-0693/) for yesterday, but not released yet?)!

Glad to hear this still made it in, thanks so much @ericsnowcurrently and @encukou!

**Comment by Yhg1s** [comment]
FWIW, this PR was one of the reasons 3.12b4 wasn't cut yesterday :)

File: Doc/c-api/typeobj.rst [hunk 5]
```diff
@@ -1717,7 +1717,19 @@ and :c:type:`PyType_Type` effectively act as defaults.)
...
+   initialization for the type has finished, this field should be
+   treated as read-only.
+
+   Some types may not store their dictionary in this slot.
+   Use :c:func:`PyType_GetDict` to retreive the dictionary for an arbitrary
+   type.
+
+   .. versionchanged:: 3.12
+
+      Internals detail: For static builtin types, this is always ``NULL``.
```
Comment (erlend-aasland) [inline Doc/c-api/typeobj.rst:1730]: Sorry for the post merge review, but FYI, there's a directive for implementation details: `.. impl-detail::`

**Comment by ericsnowcurrently** [comment]
Thanks for wrapping this up, @encukou!

**Pull request merged**